http/src/main/scala/org/http4s/blaze/http/http2/HeaderAggregatingFrameHandler.scala

```diff
@@ -0,0 +1,173 @@
+package org.http4s.blaze.http.http2
+
+import java.nio.ByteBuffer
+
+import org.http4s.blaze.http.Headers
+import org.http4s.blaze.util.BufferTools
+import Http2Exception.{PROTOCOL_ERROR, REFUSED_STREAM}
+
+/** A [[Http20FrameHandler]] that decodes raw HEADERS, PUSH_PROMISE,
+  * and CONTINUATION frames from ByteBuffer packets to a complete
+  * collections of headers.
+  *
+  * @note This class is not 'thread safe' and should be treated accordingly
+  */
+private[http2] abstract class HeaderAggregatingFrameHandler(
+    inboundSettings: Http2Settings,
+    headerDecoder: HeaderDecoder) extends Http20FrameHandler {
+
+  private[this] sealed trait PartialFrame {
+    def streamId: Int
+    var buffer: ByteBuffer
+  }
+
+  private[this] case class PHeaders(
+      streamId: Int,
+      priority: Option[Priority],
+      endStream: Boolean,
+      var buffer: ByteBuffer) extends PartialFrame
+
+  private[this] case class PPromise(
+      streamId: Int,
+      promisedId: Int,
+      var buffer: ByteBuffer) extends PartialFrame
+
+  private[this] var hInfo: PartialFrame = null
+
+
+  ///////////////////////////////////////////////////////////////////////////
+
+  /** Called on the successful receipt of a complete HEADERS block
+    *
+    * @param streamId stream id of the HEADERS block. The codec will never pass 0.
+    * @param priority optional priority information associated with this HEADERS block.
+    * @param endStream this is the last inbound frame for this stream.
+    * @param headers decompressed headers.
+    */
+  def onCompleteHeadersFrame(streamId: Int,
+                             priority: Option[Priority],
+                             endStream: Boolean,
+                             headers: Headers): Http2Result
+
+  /** Called on the successful receipt of a complete PUSH_PROMISE block
+    *
+    * @param streamId stream id of the associated stream. The codec will never pass 0.
+    * @param promisedId promised stream id. This must be a valid, idle stream id. The codec will never pass 0.
+    * @param headers decompressed headers.
+    */
+  def onCompletePushPromiseFrame(streamId: Int, promisedId: Int, headers: Headers): Http2Result
+
+  ////////////////////////////////////////////////////////////////////////////
+
+  final def setMaxHeaderTableSize(maxSize: Int): Unit = { headerDecoder.setMaxHeaderTableSize(maxSize) }
+
+  final override def inHeaderSequence(): Boolean = hInfo != null
+
+  final override def onHeadersFrame(streamId: Int,
+                                    priority: Option[Priority],
+                                    endHeaders: Boolean,
+                                    endStream: Boolean,
+                                    buffer: ByteBuffer): Http2Result = {
+
+    if (inHeaderSequence()) {
+      return PROTOCOL_ERROR.goaway(s"Received HEADERS frame while in in headers sequence. Stream $streamId").toError
+    }
+
+    if (buffer.remaining() > inboundSettings.maxHeaderListSize) {
```

Comment:
TODO: the `maxHeaderListSize` is for uncompressed headers. This should always be smaller than or at least approximately the same size as the uncompressed size, so maybe its ok?